File: cricket/src/problem/LFUCache.java
package problem;
import java.util.*;

public class LFUCache {

    // Inner class representing a custom LRUCache with LFU behavior
    class LRUCache {

        private int size; // Maximum capacity of the cache
        private Map<Integer, Integer> cache;       // Stores key -> value
        private Map<Integer, Integer> freqorder;   // Stores key -> frequency of use

        public LRUCache(int size) {
            this.size = size;
            this.cache = new HashMap<>();
            this.freqorder = new TreeMap<>();
        }

        // Get the value for the key, if it exists. Also update its frequency.
        public int get(int key) {
            if (!cache.containsKey(key)) {
                return -1;
            }

            // Increase usage frequency
            freqorder.put(key, freqorder.getOrDefault(key, 0) + 1);
            return cache.get(key);
        }

        // Insert or update the key-value pair in the cache
        public void put(int key, int value) {
            // If key already exists, just update the value and frequency
            if (cache.containsKey(key)) {
                cache.put(key, value);
                freqorder.put(key, freqorder.getOrDefault(key, 0) + 1);
                return;
            }

            // If cache is full and key doesn't already exist, evict the LFU key
            if (cache.size() >= size) {
                int lfu = getLFUKey(); // Find least frequently used key
                cache.remove(lfu);     // Remove it from cache
                freqorder.remove(lfu); // Remove it from frequency map
            }

            // Insert the new key-value pair and set its frequency to 1
            cache.put(key, value);
            freqorder.put(key, 1);
        }

        // Helper function to find the key with the least frequency
        // If tie, returns the smallest key
        private int getLFUKey() {
            int minFreq = Collections.min(freqorder.values());
            int lfuKey = Integer.MAX_VALUE;

            for (Map.Entry<Integer, Integer> entry : freqorder.entrySet()) {
                if (entry.getValue() == minFreq) {
                    lfuKey = Math.min(lfuKey, entry.getKey());
                }
            }

            return lfuKey;
        }
    }

    // Static method to process operations on the LFU cache
    public static void solve(int N, int Q, List<int[]> operations) {
        LFUCache outer = new LFUCache();
        LRUCache cache = outer.new LRUCache(N); // Create cache with capacity N

        List<Integer> result = new ArrayList<>();

        // Process each operation
        for (int[] op : operations) {
            if (op[0] == 1) {
                // Type 1: Get the value of key
                int val = cache.get(op[1]);
                result.add(val);
            } else if (op[0] == 2) {
                // Type 2: Put the key-value pair
                cache.put(op[1], op[2]);
            }
        }

        // Print all results of get operations
        for (int i = 0; i < result.size(); i++) {
            System.out.print(result.get(i));
            if (i < result.size() - 1) System.out.print(" ");
        }
        System.out.println(); // End with newline
    }

    // Sample main method for testing
    public static void main(String[] args) {
        int N = 2; // Cache capacity
        int Q = 5; // Number of operations

        // Operations in format [type, key, value or -1]
        List<int[]> operations = new ArrayList<>();
        operations.add(new int[]{1, 2, -1});  // Get 2 -> -1 (not found)
        operations.add(new int[]{2, 1, 3});   // Put 1 -> 3
        operations.add(new int[]{2, 2, 4});   // Put 2 -> 4
        operations.add(new int[]{2, 4, 5});   // Put 4 -> 5 (evict LFU key 1)
        operations.add(new int[]{1, 2, -1});  // Get 2 -> 4

        // Call the solve method to execute operations
        solve(N, Q, operations);  // Expected Output: -1 4
    }
}

File: Pyhton/Problems/LFUcache.py
class LFU:
    def __init__(self, size):
        self.size = size
        self.cache = {}      # key -> value
        self.freq = {}       # key -> frequency

    def get(self, key):
        if key in self.cache:
            self.freq[key] += 1
            return self.cache[key]
        return -1  # Not found

    def put(self, key, value):
        if self.size <= 0:
            return

        # If key exists, update value and frequency
        if key in self.cache:
            self.cache[key] = value
            self.freq[key] += 1
            return

        # If cache is full, remove the LFU entry
        if len(self.cache) >= self.size:
            min_freq = min(self.freq.values())
            lfu_keys = [k for k, f in self.freq.items() if f == min_freq]
            # Remove the oldest key among LFU keys (could use OrderedDict for true LRU)
            del_key = lfu_keys[0]
            del self.cache[del_key]
            del self.freq[del_key]

        # Insert new key
        self.cache[key] = value
        self.freq[key] = 1

def test_LFU():
    # Create LFU cache with capacity 2
    lfu = LFU(2)

    # Test 1: Put two items and get them
    lfu.put(1, 'A')   # cache: {1:'A'}
    lfu.put(2, 'B')   # cache: {1:'A', 2:'B'}
    print("Test 1.1:", "Pass" if lfu.get(1) == 'A' else "Fail")   # returns 'A'
    print("Test 1.2:", "Pass" if lfu.get(2) == 'B' else "Fail")   # returns 'B'

    # Test 2: Frequency update and eviction
    lfu.get(1)         # freq[1] = 2 now
    lfu.put(3, 'C')    # Evict key 2 (freq 1), cache: {1:'A', 3:'C'}
    print("Test 2.1:", "Pass" if lfu.get(2) == -1 else "Fail")    # returns -1 (not found)
    print("Test 2.2:", "Pass" if lfu.get(3) == 'C' else "Fail")   # returns 'C'
    print("Test 2.3:", "Pass" if lfu.get(1) == 'A' else "Fail")   # returns 'A'

    # Test 3: Further eviction
    lfu.put(4, 'D')    # Evict key 3 (freq 1), cache: {1:'A', 4:'D'}
    print("Test 3.1:", "Pass" if lfu.get(3) == -1 else "Fail")    # returns -1 (not found)
    print("Test 3.2:", "Pass" if lfu.get(4) == 'D' else "Fail")   # returns 'D'
    print("Test 3.3:", "Pass" if lfu.get(1) == 'A' else "Fail")   # returns 'A'

    print("All tests complete.")

test_LFU()